processor/src/lib.rs

```diff
@@ -93,12 +93,15 @@ pub struct ChipletsTrace {
 
 /// Returns an execution trace resulting from executing the provided program against the provided
 /// inputs.
-pub fn execute(
+pub fn execute<A>(
     program: &Program,
     stack_inputs: StackInputs,
-    advice_inputs: &ProgramInputs,
-) -> Result<ExecutionTrace, ExecutionError> {
-    let mut process = Process::new(program.kernel(), stack_inputs, advice_inputs.clone());
+    advice_provider: A,
+) -> Result<ExecutionTrace, ExecutionError>
+where
+    A: AdviceProvider,
```

Comment:
Question: why do we pass the advice provider here as owned? Wouldn't `&mut A` be a more flexible approach?

Reply:
We took here a similar approach to 

https://github.com/rust-lang/rust/blob/a5bfc25c93d2549887848529382892f93c95207d/library/std/src/io/impls.rs#L17

here

https://github.com/0xPolygonMiden/miden-vm/blob/8ce76a2e2ece12c988cf2dfa359a3def05b1df41/processor/src/advice/mod.rs#L169-L172

I like this approach because it will allow you to call `execute` passing the advice provider either as owned, or as mutable reference since both implements `AdviceProvider`

One example is [BufReader::new](https://doc.rust-lang.org/std/io/struct.BufReader.html#method.new) that takes a reader as owned, and often is used with an inner reference to actual reader (for instance, if you want a reader over a [Read::bytes](https://doc.rust-lang.org/std/io/trait.Read.html#method.bytes) iterator, then you just use `by_ref`

Reply:
Ah - so, we'd still be about to do something like this?:
```Rust
let mut advice_provider = MyAdviceProvider::new();
let result = execute(&program, stack_inputs, &mut advice_provider);
```

Reply:
Yes, this would work as well